missioncontrol/home/migrations/0021_s3files_20190227_0220.py

```diff
@@ -0,0 +1,48 @@
+# Generated by Django 2.1.4 on 2019-02-27 03:37
+
+from django.db import migrations, models
+import django.db.models.deletion
+import django.utils.timezone
+import home.models
+import uuid
+
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('home', '0020_merge_20190124_0028'),
+    ]
+
+    operations = [
+        migrations.CreateModel(
+            name='S3File',
+            fields=[
+                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
+                ('cid', models.CharField(max_length=128, unique=True)),
+                ('bucket', models.CharField(max_length=64)),
+                ('what', models.TextField(default='default')),
+                ('created', models.DateTimeField(default=django.utils.timezone.now)),
+                ('where', models.TextField(default='nowhere')),
+            ],
+            bases=(models.Model, home.models.Serializable),
+        ),
+        migrations.CreateModel(
+            name='TaskRun',
+            fields=[
+                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
+                ('uuid', models.UUIDField(default=uuid.uuid4, unique=True)),
```

Comment:
What are the implications of putting `unique=True` here?
In practice, task runs will be PUT by a client leading to an upsert. Wouldn't this only protect the case where you call create?

Reply:
Yes... but it prevents (at a database level) adding a new row with the same UUID.
There's really not a negative AFAIK.

Really should probably actually set the UUID to be the primary key....

Reply:
I read up on UUID primary keys a while back and there were a few compelling cases I read about where it screwed them down the line. I can't remember the details at this point though :/
Having our own internal ID column is definitely something I did on purpose though (as a takeaway from that research).

I don't think we need to worry about write performance at this point, I just wanted to think it though. It sounds like it will add an additional btree index for the unique constraint. :man_shrugging: